src/tir/transforms/lower_warp_memory.cc

```diff
@@ -213,9 +213,13 @@ class WarpAccessRewriter : protected StmtExprMutator {
     alloc_size *= op->dtype.lanes();
     std::tie(warp_index_, width_) = WarpIndexFinder(warp_size_).Find(op->body);
     warp_coeff_ = WarpStoreCoeffFinder(buffer_, warp_index_, analyzer_).Find(op->body);
-    CHECK_EQ(alloc_size % (width_ * warp_coeff_), 0)
-        << "Warp memory must be multiple of the extent of threadIdx.x";
-    warp_group_ = alloc_size / (width_ * warp_coeff_);
+
+    // Align the local memory size. The number of elements may not
+    // be a multiple of width_ * warp_coeff_; round it up.
+    int factor = width_ * warp_coeff_;
+    warp_group_ = (alloc_size + (factor - 1)) / factor;
+    alloc_size = warp_group_ * factor;
+
```

Comment:
Please consider this scenario:

```
// bound to threadIdx.y
for (i = 0; i < 2; i++) {
  // bound to threadIdx.x
  for (j = 0; j < 5; j++) {
    access buffer.warp[(j + 1) % 5]
  }
}
```

Now we are shuffling among Thread 0\~4, and Thread 5\~9. Since we are rounding them up to Thread 0\~7 and Thread **5\~12 (seems not right as well)**, will there be any conflict between these two group of threads?

Reply:
What’s the extent of threadidx.x? I think the test case in this PR clearly shows rounding up to the warp size is needed. It is the same as softmax failures I have seen. As there is no warp level allocation of size n,  we allocate n/32 elements in each thread. If n is not a multiple of 32, we need to ”over-allocate“ slightly and predict the access.

Reply:
> What’s the extent of threadidx.x?

It's 5. Thread 0\~4 are shuffled as a group (say Group 0), and Thread 5\~12 are another group (say Group 1). If we round up 0\~4 to 0\~7, Thread 5\~7 may be both accessed by Group 0 and Group 1, which may cause some problems.

> I think the test case in this PR clearly shows rounding up to the warp size is needed.

Yes, it is needed, but maybe not enough.

> As there is no warp level allocation of size n, we allocate n/32 elements in each thread. If n is not a multiple of 32, we need to ”over-allocate“ slightly and predict the access.

Yes, we need to over-allocate the **buffers**, but should we also over-allocate the **threads**? If not, maybe we should properly set the mask of `__shlf_sync`.

Reply:
Looks like this is a sub-warp related, and I am not following it closely. CUDA __shfl_sync does not support non-power of 2 width. So your example may not be supported.  It would be helpful if we have a legal example that would be broken by this patch. Thanks!

Reply:
Sorry I misunderstood these lines. I thought it was dealing with sub-warp logic. Then there will be no problem.

Reply:
Yes, it is not sub-warp specific. Thanks!